SFR-906 Add Edition Detail Endpoint

mwbenowitz opened the pull request:
Add a new endpoint that accepts an edition row id and returns a response detailing the edition and its component instances. This enables users to see the source data that was "clustered" into the edition.

The response includes metadata from the edition and an array of instances sorted by the number of holdings reported by OCLC, with the first item being a special case in that it is always the instance that contains the links made available through the edition.

mwbenowitz committed:
SFR-906 Add Edition detail endpoint

Add a new endpoint that accepts an edition row id and returns a response detailing the edition and its component instances. This enables users to see the source data that was "clustered" into the edition.

The response includes metadata from the edition and an array of instances sorted by the number of holdings reported by OCLC, with the first item being a special case in that it is always the instance that contains the links made available through the edition.

mwbenowitz committed:
SFR-906 Update Edition Endpoint to fix bugs

This fixes a few minor bugs revealed in testing:
- Returns clear error message if edition not found
- Properly sorts instances when a featured instance is present

mwbenowitz committed:
SFR-906 Add unit and integration tests for v3Edition endpoint

mwbenowitz committed:
SFR-906 Add documentation, including Swagger and CHANGELOG

hokei commented on CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 This file documents all updates and releases to the ResearchNow Data Ingest pipeline.
 
+## [0.0.5] - Unreleased
+### Added
```
I see that the ticket about adding edition identifiers to search results and work details is closed - was that done here or was it already available in the api?

mwbenowitz replied:
It's already available in the API, each `edition` object has an `id` attribute. I think it was included originally to help with React `key`s, but it is the right identifier to hit the endpoint

hokei commented on app/sfr-search-api/lib/v3Edition.js:
```diff
@@ -0,0 +1,140 @@
+    this.sortInstances()
+
+    // Remove items and covers from the edition; these are displayed on instances
+    delete this.edition.items
```
I think we go through this over and over again but ... you're gonna have to explain to me the difference between items and instances again sorry orz.  

Each card inside the edition details page represents an instance, right?  But don't we iterate through "item.links" to get the read online/download links ... ?

hokei replied:
Oh I see below, so you have `instance.items`.  got it.

mwbenowitz replied:
yeah, in the database items (and their links) are always related to instances. For the work detail we pull out one item and add it to the edition for the card.

This reuses some of that work detail code, so we just have to remove that item from the edition since for the work detail all items are attached to instances

hokei commented on app/sfr-search-api/lib/v3Search.js:
```diff
@@ -188,7 +188,7 @@ class V3Search {
    */
   async getInstances(instanceIds) {
-    return this.dbConn.loadInstances(instanceIds, 1)
+    return this.dbConn.loadInstances(instanceIds, 1, ['agents', 'items', 'languages', 'covers', 'dates', 'rights'])
```
I'd assign this to a constant somewhere easily editable and commented.  (Should this always be the same value as in `v3Work.js` below?

mwbenowitz replied:
A good idea, and yeah it should be the same as the work detail. There are probably other constants that could be refactored in that way too.

hokei commented on app/sfr-search-api/swagger.v3.json:
```diff
@@ -1400,6 +1519,71 @@
+          "type": "string",
+          "format": "date"
+        },
+        "edition": {
```
is this edition id or something different?

mwbenowitz replied:
This is the edition description ("2nd ed" etc.), the `edition_id` is the `id` field a few places up from this.

mwbenowitz committed:
SFR-906 Refactor join table lists to separate file

This separates out the arrays of tables to be joined to the main query via subqueries into a separate helper module. There they can be updated for all places they are used, and reducing some duplication in the code.

mwbenowitz committed:
SFR-906 Add showAll parameter to edition detail endpoint

This parameter controls which instance records are included in the edition detail endpoint. By default the endpoint returns all instances. If `showAll` is set to "false" only istances that have read online/download links are included are included. This enables filtering toggles on the detail page in the React app to function in a no-js environment
